"""
ALPACA and PEANUT based SNV calling workflow.

Expects a json config file with the following structure, assuming that the
desired reference sequence is some genome
to be found under the given path, and two units A and B have been sequenced with Illumina.

{
    "references": {
        "genome": "path/to/genome.fasta"
    },
    "samples": {
        "A": ["A"],
        "B": ["B"]
    },
    "units": {
        "A":
            ["path/to/A_R1.fastq.gz", "path/to/A_R2.fastq.gz"],
        "B":
            ["path/to/B.fastq.gz"]
    },
    "known_variants": {
        "dbsnp": "path/to/dbsnp.vcf"
    },
    "platform": "Illumina",
    "ploidy": 2,
    "heterozygosity": 0.001,
    "alpaca_queries": {
        "somatic": "A - B"
    },
    "min_qual": 200,
    "max_strandbias": 50
}

Note the separation between samples and units that allows to have more than
one sequencing run for each sample, or multiple lanes per sample.
"""


__author__ = "Johannes KÃ¶ster"
__license__ = "MIT"


import json


with open("config.json") as f:
    CONFIG = json.load(f)


include:
    "https://bitbucket.org/johanneskoester/snakemake-workflows/raw/master/bio/ngs/rules/mapping/peanut.rules"
include: 
    "https://bitbucket.org/johanneskoester/snakemake-workflows/raw/master/bio/ngs/rules/mapping/samfiles.rules"
include:
    "https://bitbucket.org/johanneskoester/snakemake-workflows/raw/master/bio/ngs/rules/snv_calling/alpaca_calling.rules"


rule all:
    input:
        expand("snv_calling/genome/{alpaca_queries}.{min_qual}.{max_strandbias}.vcf", **CONFIG)
